client-lib/nsa.js

```diff
@@ -0,0 +1,210 @@
+  Object.keys(playerMap).forEach((k) => {
+    if (playerMap[k]) playerMap[k].pause();
+  });
+
+  deepAssign(window.AppData, {
+    currentTime: 0,
+    playing: false,
+    error: false,
+    exited: false,
+    time: `${formatTime(0)} / ${formatTime(window.AppData.duration)}`,
+    progress: 0.001
+  });
+})
+
+emitter.on('init', (opts) => {
+  sendMetricsEvent('player_view', 'init');
+
+  if (window.AppData.player === 'audio') {
+    playerMap['audio'] = new AudioCtrl(opts);
+  } else if (window.AppData.player === 'youtube') {
+    const PLAYING = window.YT.PlayerState.PLAYING;
+    const PAUSED = window.YT.PlayerState.PAUSED;
+
+    playerMap['youtube'] = new YtCtrl({
+      id: 'video-yt',
+      onReady: (ev) => {opts.onLoaded(ev.target.getDuration())},
+      onStateChange: (ev) =>  {
+        if (ev.data === PLAYING && !window.AppData.playing) emitter.emit('play')
+        else if (ev.data === PAUSED && window.AppData.playing) emitter.emit('pause')
+      },
+      onError: (err) => window.AppData.error = true
+    });
+
+  } else {
+    playerMap['video'] = new VideoCtrl(opts);
+  }
+});
+
+emitter.on('play', (opts) => {
+  sendMetricsEvent('player_view', 'play');
+  // TODO(DJ): this is a special case to handle audio replays this is
+  // a work around for a bug in the audiosource dependency
+  if (opts && opts.replay) {
+    playerMap[window.AppData.player].play(0.001);
+  } else playerMap[window.AppData.player].play();
+
+  window.AppData.playing = true;
+  currentStep = requestAnimationFrame(step);
+});
+
+emitter.on('pause', (opts) => {
+  sendMetricsEvent('player_view', 'pause');
+  playerMap[window.AppData.player].pause();
+  window.AppData.playing = false;
+});
+
+emitter.on('mute', (opts) => {
+  sendMetricsEvent('player_view', 'mute');
+  playerMap[window.AppData.player].mute();
+  window.AppData = deepAssign(window.AppData, {muted: true});
+});
+
+emitter.on('unmute', (opts) => {
+  sendMetricsEvent('player_view', 'unmute');
+  playerMap[window.AppData.player].unmute();
+  window.AppData = deepAssign(window.AppData, {muted: false});
+});
+
+emitter.on('replay', (opts) => {
+  sendMetricsEvent('player_view', 'replay');
+  emitter.emit('reset');
+  emitter.emit('play', {replay: true});
+});
+
+emitter.on('load', (opts) => {
+  sendMetricsEvent('player_view', 'video_loaded');
+
+  // initial step to set times
+  step();
+
+  window.AppData = deepAssign(window.AppData, {
+    loaded: true,
+    duration: opts.duration
+  });
+});
+
+emitter.on('set-volume', (opts) => {
+  playerMap[window.AppData.player].volume = opts.value;
+  window.AppData = deepAssign(window.AppData, {
+    volume: opts.value
+  });
+});
+
+emitter.on('set-time', (opts) => {
+  playerMap[window.AppData.player].time = opts.value;
+
+  // if we are paused force the ui to update
+  if (!window.AppData.playing) {
+    window.AppData = deepAssign(window.AppData, {
+      time: `${formatTime(opts.value)} / ${formatTime(window.AppData.duration)}`,
```

Comment:
hmm. it might also make sense to split this time thing into 'elapsed time' and 'total time'. that would let you do the locale-specific display stuff in the display code, where it belongs, and not in the data management code.